File: BOJ/BOJ_1976.cpp
#include <cstdio>
#include <iostream>
#include <algorithm>
#include <string>
#include <string.h>
#include <vector>
#include <queue>
#include <set>
#include <map>
#include <climits>


using namespace std;

int parent[201];

//union-find

int find(int n) {
	if (parent[n] < 0) return n;
	parent[n] = find(parent[n]);
	return parent[n];
}
void merge(int a, int b) {
	a = find(a);
	b = find(b);
	if (a == b) return;
	parent[b] = a;
}

int main()
{
	int n, m;
	scanf("%d %d", &n, &m);
	int connect;
	for (int i = 0; i < n; i++) {
		parent[i] = -1;
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			scanf("%d", &connect);
			if (connect == 1) {
				merge(i, j);
			}
		}
	}
	int st, go;
	scanf("%d", &st);
	bool isPossible = true;
	for (int i = 0; i < m - 1; i++) {
		scanf("%d", &go);
		if (find(st) != find(go)) {
			isPossible = false;
			break;
		}
	}
	if (isPossible) {
		printf("YES\n");
	}
	else {
		printf("NO\n");
	}
	return 0;
}

File: BOJ/BOJ_1976.py


import sys


def read():
    return sys.stdin.readline().rstrip()


def find(u):
    if u == parent[u]:
        return u
    parent[u] = find(parent[u])
    return parent[u]


def union(u, v):
    u = find(u)
    v = find(v)
    if u == v:
        return
    parent[v] = u


N = int(read())
M = int(read())
g = []
for _ in range(N):
    g.append(list(map(int, read().split())))
plan = list(map(lambda x: int(x)-1, read().split()))

parent = [i for i in range(N)]
for i in range(N):
    for j in range(N):
        if g[i][j] == 1:
            union(i, j)

answer = "YES"
start = plan[0]
for i in range(1, M):
    if find(start) != find(plan[i]):
        answer = "NO"
        break
print(answer)
